tests/integration/ha/test_ha.py

```diff
@@ -0,0 +1,337 @@
+    )
+    await send_control_signal(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{leader_num}", signal="SIGTERM"
+    )
+    # Give time for the service to restart
+    await asyncio.sleep(REELECTION_TIME * 2)
+
+    initial_offsets = get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    await asyncio.sleep(CLIENT_TIMEOUT)
+    next_offsets = get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+
+    topic_description = get_topic_description(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    assert topic_description.in_sync_replicas == {0, 1, 2}
+    assert int(next_offsets[-1]) > int(initial_offsets[-1])
+
+    result = c_writes.stop()
+    assert_continuous_writes_consistency(result=result)
+
+
+async def test_freeze_broker_with_topic_leader(
+    ops_test: OpsTest,
+    c_writes: ContinuousWrites,
+    c_writes_runner: ContinuousWrites,
+):
+    topic_description = get_topic_description(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    initial_leader_num = topic_description.leader
+
+    logger.info(
+        f"Freezing broker of leader for topic '{ContinuousWrites.TOPIC_NAME}': {initial_leader_num}"
+    )
+    await send_control_signal(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{initial_leader_num}", signal="SIGSTOP"
+    )
+    await asyncio.sleep(REELECTION_TIME * 2)
+
+    # verify replica is not in sync
+    topic_description = get_topic_description(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    assert topic_description.in_sync_replicas == {0, 1, 2} - {initial_leader_num}
+    assert initial_leader_num != topic_description.leader
+
+    # verify new writes are continuing. Also, check that leader changed
+    topic_description = get_topic_description(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    initial_offsets = get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+    await asyncio.sleep(CLIENT_TIMEOUT * 2)
+    next_offsets = get_topic_offsets(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+
+    assert int(next_offsets[-1]) > int(initial_offsets[-1])
+
+    # Un-freeze the process
+    logger.info(f"Un-freezing broker: {initial_leader_num}")
+    await send_control_signal(
+        ops_test=ops_test, unit_name=f"{APP_NAME}/{initial_leader_num}", signal="SIGCONT"
+    )
+    await asyncio.sleep(REELECTION_TIME)
+    topic_description = get_topic_description(ops_test=ops_test, topic=ContinuousWrites.TOPIC_NAME)
+
+    # verify the unit is now rejoined the cluster
+    assert topic_description.in_sync_replicas == {0, 1, 2}
+
+    result = c_writes.stop()
+    assert_continuous_writes_consistency(result=result)
+
+
+async def test_full_cluster_crash(
+    ops_test: OpsTest,
+    restart_delay,
+    c_writes: ContinuousWrites,
+    c_writes_runner: ContinuousWrites,
+):
+    # Let some time pass for messages to be produced
+    await asyncio.sleep(10)
```

Comment:
**nitpick:** Could we use a global var here like with other tests?

Reply:
Same as [this](https://github.com/canonical/kafka-k8s-operator/pull/68#discussion_r1365387183). I'll note this down and update on both charms to have more consistent waits.